kimchi/src/snarky/ec.rs

```diff
@@ -0,0 +1,304 @@
+                    F::from(1_u8)
+                } else {
+                    F::zero()
+                }
+            };
+            let x1 = w.read_var(&a0);
+            let x2 = w.read_var(&a1);
+            let y1 = w.read_var(&b0);
+            let y2 = w.read_var(&b1);
+
+            let same_x = to_bool(x1 == x2);
+
+            let x_diff = x2 - x1;
+            let x21_inv = x_diff.inverse().unwrap();
+            let y_diff = y2 - y1;
+            let slope = y_diff * x21_inv;
+
+            let a = GroupAffine::<C>::new(x1, y1, false);
+            let b = GroupAffine::<C>::new(x2, y2, false);
+            assert!(!a.infinity, "can not be infinity");
+            assert!(!b.infinity, "can not be infinity");
+
+            let c = a + b;
+            let inf = to_bool(c.infinity);
+
+            let inf_z = if y1 == y2 {
+                F::zero()
+            } else if x1 == x2 {
+                y_diff.inverse().unwrap()
+            } else {
+                F::zero()
+            };
+            [c.x, c.y, inf, same_x, slope, inf_z, x21_inv]
+        })?;
+    let p1 = (a0, a1);
+    let p2 = (b0, b1);
+    let p3 = (p3x.clone(), p3y.clone());
+    let c = [p3x, p3y];
+    inf.assert_equals(runner, loc.clone(), &FieldVar::zero())?;
+
+    let constraint =
+        Constraint::KimchiConstraint(KimchiConstraint::EcAddComplete(EcAddCompleteInput {
+            p1,
+            p2,
+            p3,
+            inf,
+            same_x,
+            slope,
+            inf_z,
+            x21_inv,
+        }));
+    runner.add_constraint(constraint, None, loc)?;
+    Ok(c)
+}
+
+pub fn ec_scale<F: PrimeField, C: SWModelParameters<BaseField = F>>(
+    runner: &mut RunState<F>,
+    loc: Cow<'static, str>,
+    base: [FieldVar<F>; 2],
+    scalar: &FieldVar<F>,
+) -> SnarkyResult<[FieldVar<F>; 2]> {
+    let limbs = split_in_limbs(runner, loc.clone(), scalar)?;
+    let limbs: [Limb<F>; 51] = limbs.into();
+    let n_prev = FieldVar::zero();
+    let state: Option<([FieldVar<F>; 2], FieldVar<F>)> = None;
+    let rounds = limbs.into_iter().scan(state, |state, scalar| {
+        let round = match state {
+            Some((input, n_prev)) => limb_scale::<F, C>(
+                runner,
+                loc.clone(),
+                base.clone(),
+                input.clone(),
+                scalar,
+                n_prev.clone(),
+            ),
+            None => limb_scale::<F, C>(
+                runner,
+                loc.clone(),
+                base.clone(),
+                base.clone(),
+                scalar,
+                n_prev.clone(),
+            ),
+        };
+        let out = round.as_ref().unwrap().0.clone();
+        let n_next = round.as_ref().unwrap().1.n_next.clone();
+        *state = Some((out, n_next));
+        let round = round.map(|r| r.1);
+        Some(round)
+    });
+    let rounds: SnarkyResult<Vec<ScaleRound<FieldVar<F>>>> = rounds.collect();
+    let rounds = rounds?;
```

Comment:
Is this explicit unwrapping (and also here https://github.com/o1-labs/proof-systems/pull/1764/files#diff-280f21b46542317fc5664ff595b11c9a91d03f9b76aa17dc9d6f96d3826dc41bR131) necessary for some type safety reasons? As opposed to just using `?` on the previous line directly.